packages/components/src/navigation/NavTabList.tsx

```diff
@@ -0,0 +1,455 @@
+      clearTimeout(holdTimer);
+      continuousScrollRef.current.holdTimer = undefined;
+    }
+    if (rAF != null) {
+      cancelAnimationFrame(rAF);
+      continuousScrollRef.current.rAF = undefined;
+    }
+    window.removeEventListener('mouseup', endContinuousScroll);
+  }, []);
+
+  useEffect(
+    () => () => window.removeEventListener('mouseup', endContinuousScroll),
+    [endContinuousScroll]
+  );
+
+  const handleMouseDown = useCallback(
+    (direction: 'left' | 'right') => {
+      if (containerRef.current != null) {
+        continuousScrollRef.current.holdTimer = window.setTimeout(
+          handleMouseRepeat,
+          CLICK_TIMEOUT,
+          direction,
+          containerRef.current.scrollLeft
+        );
+      }
+      continuousScrollRef.current.cancelClick = false;
+      window.addEventListener('mouseup', endContinuousScroll);
+    },
+    [endContinuousScroll, handleMouseRepeat]
+  );
+
+  const handleMouseDownLeft = useCallback(() => {
+    handleMouseDown('left');
+  }, [handleMouseDown]);
+
+  const handleMouseDownRight = useCallback(() => {
+    handleMouseDown('right');
+  }, [handleMouseDown]);
+
+  // React binds to the root as a passive listener for wheel
+  // This prevents the wheel event from being canceled
+  // Bypass React's event system so we can prevent the default behavior
+  // https://github.com/facebook/react/issues/14856
+  useEffect(function handleWheel() {
+    const onWheel = (e: WheelEvent) => {
+      e.stopPropagation();
+      e.preventDefault();
+      const nav = e.currentTarget as HTMLDivElement;
+      const delta =
+        Math.abs(e.deltaY) > Math.abs(e.deltaX) ? e.deltaY : e.deltaX;
+
+      // Scrolling jumps too far sometimes, so clamp to get a smoother scroll
+      nav.scrollLeft += clamp(delta, -30, 30);
+    };
+
+    containerRef.current?.addEventListener('wheel', onWheel);
+    return () => {
+      containerRef.current?.removeEventListener('wheel', onWheel);
+    };
+  }, []);
+
+  const tabContextActionMap = useMemo(() => {
+    const tabContextActions = new Map<string, ResolvableContextAction[]>();
+    tabs.forEach(tab => {
+      const { key } = tab;
+      const contextActions = [
+        ...makeBaseContextItems(tab, tabs, onClose),
+        ...(makeContextItems?.(tab) ?? []),
+      ];
+      tabContextActions.set(key, contextActions);
+    });
+    return tabContextActions;
+  }, [makeContextItems, tabs, onClose]);
+
+  const activeTabRef = useRef<HTMLDivElement>(null);
+  const activeTab = tabs.find(tab => tab.key === activeKey) ?? tabs[0];
+  const navTabs = tabs.map((tab, index) => {
+    const { key } = tab;
+    const isActive = tab === activeTab;
+
+    return (
+      <NavTab
+        tab={tab}
+        key={key}
+        index={index}
+        isActive={isActive}
+        activeRef={activeTabRef}
+        onSelect={onSelect}
+        onClose={onClose}
+        isDraggable={isReorderAllowed}
+        contextActions={tabContextActionMap.get(key)}
```

Comment:
Instead of making all the context actions ahead of time, you should be passing a `ResolvableContextAction`, which can just be a function to get the context actions for that tab item, and then they generate on demand when the item is right-clicked.
And no need to join them, context actions already handles that. So you can pass:
```
contextActions={[() => makeBaseContextItems(tab, tabs, onClose), () => (makeContextItems?.(tab) ?? [])}
```
Or you can just do that in the `tabContextActionMap` above, so it'd be:
```
      const contextActions = [
        () => makeBaseContextItems(tab, tabs, onClose),
        () => (makeContextItems?.(tab) ?? []),
      ];
```

Reply:
Ya good point. They're generated ahead of time on DHE and I just kept that behavior